### ISSUE TITLE
marshal-mongo query with $and not generated correctly

### ISSUE THREAD

**Issue by colorcube** [super-hornet/super-hornet.ts#61]
This is related to @marcj/marshal-mongo@2.1.13
While the original repository is no longer online and this is the successor I place it here.

Using the filter() method results in a wrong query when using $and:

original filter: `{ tags: { '$all': [ 'one', 'two' ] } }`
processed filter:  `{ tags: { '$all': [ [], [] ] } }`

Here's example code
I use convertPlainQueryToMongo directly here but this will also happen with filter() (which calls convertPlainQueryToMongo!?)

```
    class Test {
        @f.array(String)
        public tags: string[] = [];
    }

    let tags = ['one', 'two'];

    let filter = {
        tags: { $all: tags }
    };

    console.log('filter plain', filter);

    const fieldNames = {};
    const m = convertPlainQueryToMongo(Test, filter, fieldNames);
    console.log('filter mongo', m);


    // filter plain (expected): { tags: { '$all': [ 'one', 'two' ] } }
    // filter mongo (wrong):    { tags: { '$all': [ [], [] ] } }
```

I tracked down the Problem in convertQueryToMongo()

```
                    else if (j === '$in' || j === '$nin' || j === '$all') {
                        fieldNamesMap[targetI] = true;
                        fieldValue[j] = queryValue.map(v => converter(classType, targetI, v));

                        // using this here generates the right query - no idea if this is bad
                        // fieldValue[j] = converter(classType, targetI, queryValue);
```

I guess this will not be fixed in marshal-mongo@2.1.13, so I have a few questions:

- am I doing something wrong?
- can I circumvent the problem?
- can I use marshal and marshall-mongo from super-hornet already or is it too early?

**Comment by marcj** [comment]
> am I doing something wrong?

No, you're using it right. Indeed it's a bug which I fixed [right away](https://github.com/super-hornet/super-hornet.ts/commit/fe0ef6b08e9c8484dd2bd0b60c2ae1e4a9bdfa55).

Your custom hotfix breaks $in/$nin, but works for the most basic usage of $all.

> can I circumvent the problem?

Unfortunately there's no workaround possible in the old version.

> can I use marshal and marshall-mongo from super-hornet already or is it too early?

Not yet, since we don't have a tagged release. But I surely could publish an alpha release. The new marshal-mongo is MUCH faster than the old one, since a custom BSON parser and Mongo client has been implemented, which of course needs first some stability checks, but test coverage is high. So if you want to try it out yet I can tag a release. But before that I'm going to move/rename this project once again to its final name.
